code/game/machinery/autolathe/matterforge.dm

```diff
@@ -0,0 +1,594 @@
+		saved_mat["req"] = total_mat
+		saved_mat["name"] = MATERIAL_COMPRESSED_MATTER
+		ui_mats = list(1)
+		ui_mats[1] = saved_mat
+		design_object.ui_data["materials"] = ui_mats
+		files.Add(design_object)
+	qdel(c)
+	return files
+
+/obj/machinery/matter_nanoforge/proc/materials_data()
+	var/list/data = list()
+	data["mat_efficiency"] = mat_efficiency
+	data["mat_capacity"] = storage_capacity
+
+	var/list/M = list()
+	for(var/mtype in stored_material)
+		if(stored_material[mtype] <= 0)
+			continue
+
+		var/material/MAT = get_material_by_name(mtype)
+		var/list/ME = list("name" = MAT.display_name, "id" = mtype, "amount" = stored_material[mtype], "ejectable" = !!MAT.stack_type)
+
+		M.Add(list(ME))
+
+	data["materials"] = M
+
+	return data
+
+/obj/machinery/matter_nanoforge/ui_data()
+	var/list/data = list()
+
+	data["have_materials"] = have_materials
+	data["have_design_selector"] = have_design_selector
+
+	data["error"] = error
+	data["paused"] = paused
+
+	data["unfolded"] = unfolded
+
+	if(categories)
+		data["categories"] = categories
+		data["show_category"] = show_category
+
+	data["special_actions"] = special_actions
+
+	data |= materials_data()
+
+	var/list/L = list()
+	for(var/datum/design/d in design_list)
+		L.Add(list(d.ui_data))
+	data["designs"] = L
+
+
+	if(current_design)
+		data["current"] = current_design.ui_data
+		data["progress"] = progress
+
+	var/list/Q = list()
+	var/qmats = stored_material[MATERIAL_COMPRESSED_MATTER]
+
+	for(var/i = 1; i <= queue.len; i++)
+		var/datum/design/picked_design = queue[i]
+		var/list/QR = picked_design.ui_data
+
+		QR["ind"] = i
+
+		QR["error"] = 0
+
+		for(var/rmat in picked_design.materials)
+			qmats -= picked_design.materials[rmat]
+			if(qmats < 0)
+				QR["error"] = 1
+
+		if(can_print(picked_design) != ERR_OK)
+			QR["error"] = 2
+
+		Q.Add(list(QR))
+
+	data["queue"] = Q
+	data["queue_max"] = queue_max
+
+	return data
+
+/obj/machinery/matter_nanoforge/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = NANOUI_FOCUS)
+	var/list/data = ui_data(user, ui_key)
+
+	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if (!ui)
+		// the ui does not exist, so we'll create a new() one
+		// for a list of parameters and their descriptions see the code docs in \code\modules\nano\nanoui.dm
+		ui = new(user, src, ui_key, "autolathe.tmpl", capitalize(name), 600, 700)
+
+		// template keys starting with _ are not appended to the UI automatically and have to be called manually
+		ui.add_template("_materials", "autolathe_materials.tmpl")
+		ui.add_template("_reagents", "autolathe_reagents.tmpl")
+		ui.add_template("_designs", "matterforge_designs.tmpl")
+		ui.add_template("_queue", "autolathe_queue.tmpl")
+
+		// when the ui is first opened this is the data it will use
+		ui.set_initial_data(data)
+		// open the new ui window
+		ui.open()
+
+/obj/machinery/matter_nanoforge/attack_hand(mob/user)
+	if(..())
+		return TRUE
+	matter_assoc_list()
```

Comment:
you don't need to go through the entire code by running this proc. You only need to declare the list "lst" from the beginning, when you have declared the object (on line 24).

Reply:
done